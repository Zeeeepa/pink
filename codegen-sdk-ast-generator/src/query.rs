use std::{collections::BTreeMap, sync::Arc};

use codegen_sdk_common::{
    CSTNode, HasChildren, Language,
    naming::{normalize_field_name, normalize_type_name},
};
use codegen_sdk_cst::CSTLanguage;
use codegen_sdk_cst_generator::{Config, Field, State};
use codegen_sdk_ts_query::cst as ts_query;
use derive_more::Debug;
use log::{debug, info, warn};
use proc_macro2::{Ident, TokenStream};
use quote::{format_ident, quote};
fn captures_for_field_definition<'a>(
    node: &ts_query::FieldDefinition<'a>,
) -> impl Iterator<Item = ts_query::Capture<'a>> {
    let mut captures = Vec::new();
    for child in node.children() {
        match child {
            ts_query::FieldDefinitionChildren::NamedNode(named) => {
                captures.extend(captures_for_named_node(&named));
            }
            ts_query::FieldDefinitionChildren::FieldDefinition(field) => {
                captures.extend(captures_for_field_definition(&field));
            }
            _ => {}
        }
    }
    captures.into_iter()
}
fn captures_for_named_node<'a>(
    node: &ts_query::NamedNode<'a>,
) -> impl Iterator<Item = ts_query::Capture<'a>> {
    let mut captures = Vec::new();
    for child in node.children() {
        match child {
            ts_query::NamedNodeChildren::Capture(capture) => captures.push(capture),
            ts_query::NamedNodeChildren::NamedNode(named) => {
                captures.extend(captures_for_named_node(&named));
            }
            ts_query::NamedNodeChildren::FieldDefinition(field) => {
                captures.extend(captures_for_field_definition(&field));
            }
            _ => {}
        }
    }
    captures.into_iter()
}
#[derive(Debug)]
pub struct Query<'a> {
    node: ts_query::NamedNode<'a>,
    language: &'a Language,
    pub(crate) state: Arc<State<'a>>,
}
impl<'a> Query<'a> {
    pub fn from_queries(
        db: &'a dyn salsa::Database,
        source: &str,
        language: &'a Language,
    ) -> BTreeMap<String, Self> {
        let parsed = ts_query::Query::parse(db, source.to_string())
            .as_ref()
            .unwrap();
        let config = Config::default();
        let state = Arc::new(State::new(language, config));
        let mut queries = BTreeMap::new();
        for node in parsed.children() {
            match node {
                ts_query::ProgramChildren::NamedNode(named) => {
                    let query = Self::from_named_node(&named, language, state.clone());
                    queries.insert(query.name(), query);
                }
                node => {
                    println!("Unhandled query: {:#?}", node);
                }
            }
        }

        // let root_node: Node<'a> = tree.root_node();
        // for child in root_node.children(&mut root_node.walk()) {
        //     if child.kind() == "grouping" {
        //         continue;
        //         // TODO: Handle grouping
        //     }
        //     let query = Self {
        //         query: child,
        //         source: source.to_string(),
        //     };
        //     queries.insert(query.name(), query);
        // }
        queries
    }
    fn from_named_node(
        named: &ts_query::NamedNode<'a>,
        language: &'a Language,
        state: Arc<State<'a>>,
    ) -> Self {
        Query {
            node: named.clone(),
            language: language,
            state: state,
        }
    }
    /// Get the kind of the query (the node to be matched)
    pub fn kind(&self) -> String {
        if let ts_query::NamedNodeName::Identifier(identifier) = &(*self.node.name) {
            return identifier.source();
        }
        panic!("No kind found for query. {:#?}", self.node);
    }
    pub fn struct_name(&self) -> String {
        normalize_type_name(&self.kind(), true)
    }
    pub fn struct_variants(&self) -> Vec<String> {
        if self
            .state
            .get_node_for_struct_name(&self.struct_name())
            .is_some()
        {
            return vec![self.struct_name()];
        }
        self.state
            .get_variants(&self.struct_name())
            .into_iter()
            .map(|v| v.normalize())
            .filter(|v| v != "Comment")
            .collect()
    }

    fn captures(&self) -> Vec<ts_query::Capture> {
        captures_for_named_node(&self.node).collect()
    }
    /// Get the name of the query (IE @reference.class)
    pub fn name(&self) -> String {
        let mut result = self.captures().last().unwrap().source();
        result.replace_range(0..1, "");
        result
    }

    //     for node in self.query.named_children(&mut self.query.walk()) {
    //         for node in node.named_children(&mut self.query.walk()) {
    //             if node.kind() == "capture" {
    //                 return get_text_from_node(&node, &self.source);
    //             }
    //         }
    //     }

    //     panic!(
    //         "No name found for query. {:?}\n{}",
    //         self.query,
    //         self.source()
    //     );
    // }

    // fn execute<T: HasChildren>(&self, node: &T) -> Vec<Box<dyn CSTNode + Send>> {
    //     let mut result = Vec::new();

    //     for child in node.children() {
    //         if self
    //             .captures()
    //             .iter()
    //             .any(|capture| capture.source() == child.kind())
    //         {
    //             result.push(child);
    //         }
    //     }
    //     result
    // }
    pub fn node(&self) -> &ts_query::NamedNode {
        &self.node
    }
    pub fn executor_id(&self) -> Ident {
        let raw_name = self.name();
        let name = raw_name.split(".").last().unwrap();
        if name.ends_with("s") {
            format_ident!("{}es", name)
        } else {
            format_ident!("{}s", name)
        }
    }
    fn get_field_for_field_name(&self, field_name: &str, struct_name: &str) -> Option<&Field> {
        debug!(
            "Getting field for: {:#?} on node: {:#?}",
            field_name, struct_name
        );
        let node = self.state.get_node_for_struct_name(struct_name);
        if let Some(node) = node {
            return node.get_field_for_field_name(field_name);
        }
        warn!(
            "No node found for: {:#?}. In language: {:#?}",
            struct_name,
            self.language.name()
        );
        None
    }
    fn get_matcher_for_field(
        &self,
        field: &ts_query::FieldDefinition,
        struct_name: &str,
        current_node: &Ident,
        name_value: Option<TokenStream>,
    ) -> TokenStream {
        let other_child: ts_query::NodeTypes =
            field.children().into_iter().skip(2).next().unwrap().into();
        for name in &field.name {
            if let ts_query::FieldDefinitionName::Identifier(identifier) = name {
                let name = normalize_field_name(&identifier.source());
                if let Some(field) = self.get_field_for_field_name(&name, struct_name) {
                    let field_name = format_ident!("{}", name);
                    let normalized_struct_name = field.type_name();
                    let wrapped = self.get_matcher_for_definition(
                        &normalized_struct_name,
                        other_child.clone(),
                        &field_name,
                        name_value,
                    );
                    assert!(
                        wrapped.to_string().len() > 0,
                        "Wrapped is empty, {} {} {}",
                        normalized_struct_name,
                        other_child.source(),
                        other_child.kind()
                    );
                    if !field.is_optional() {
                        return quote! {
                            let #field_name = &*#current_node.#field_name;
                            #wrapped
                        };
                    } else {
                        return quote! {
                            if let Some(#field_name) = &*#current_node.#field_name {
                                #wrapped
                            }
                        };
                    }
                } else {
                    panic!(
                        "No field found for: {:#?} on node: {:#?}. In language: {:#?}. Field source: {:#?}",
                        name,
                        struct_name,
                        self.language.name(),
                        field.source()
                    )
                }
            }
        }
        panic!(
            "No field found for: {:#?}. In language: {:#?}",
            field.source(),
            self.language.name()
        )
    }
    fn get_matchers_for_grouping(
        &self,
        node: &ts_query::Grouping,
        struct_name: &str,
        current_node: &Ident,
        name_value: Option<TokenStream>,
    ) -> TokenStream {
        let mut matchers = TokenStream::new();
        for group in node.children() {
            let result = self.get_matcher_for_definition(
                struct_name,
                group.into(),
                current_node,
                name_value.clone(),
            );
            matchers.extend_one(result);
        }
        matchers
    }
    fn _get_matcher_for_named_node(
        &self,
        struct_name: &str,
        target_name: &str,
        target_kind: &str,
        current_node: &Ident,
        remaining_nodes: Vec<ts_query::NamedNodeChildren<'_>>,
        name_value: Option<TokenStream>,
    ) -> TokenStream {
        let mut matchers = TokenStream::new();
        let mut field_matchers = TokenStream::new();
        let mut comment_variant = None;
        let variants = self.state.get_variants(&format!("{}Children", target_kind));
        if variants.len() == 2 {
            if variants.iter().any(|v| v.normalize() == "Comment") {
                for variant in variants {
                    if variant.normalize() == "Comment" {
                        continue;
                    }
                    comment_variant = Some(variant.normalize());
                }
            }
        }

        for child in remaining_nodes {
            if child.kind() == "field_definition" {
                field_matchers.extend_one(self.get_matcher_for_definition(
                    &target_name,
                    child.into(),
                    current_node,
                    name_value.clone(),
                ));
            } else {
                let result = self.get_matcher_for_definition(
                    &target_name,
                    child.into(),
                    &format_ident!("child"),
                    name_value.clone(),
                );

                if let Some(ref variant) = comment_variant {
                    let children = format_ident!("{}Children", target_name);
                    let variant = format_ident!("{}", variant);
                    matchers.extend_one(quote! {

                        if let crate::cst::#children::#variant(#current_node) = #current_node {
                            #result
                        }
                    });
                } else {
                    matchers.extend_one(quote! {
                        #result
                    });
                }
            }
        }
        let matchers = if matchers.is_empty() {
            quote! {}
        } else {
            quote! {
                for child in #current_node.children().into_iter() {
                    #matchers
                    break;
                }
            }
        };
        let query_source = format!(
            "Code for query: {}",
            &self.node().source().replace("\n", " ") // Newlines mess with quote's doc comments
        );
        if matchers.is_empty() && field_matchers.is_empty() {
            return quote! {};
        }
        let base_matcher = quote! {
            #[doc = #query_source]
            #matchers
            #field_matchers
        };
        if struct_name == target_name {
            return base_matcher;
        } else {
            let mut children = format_ident!("{}", struct_name);
            if let Some(node) = self.state.get_node_for_struct_name(struct_name) {
                children = format_ident!("{}", node.children_struct_name());
            }
            let variant = format_ident!("{}", target_name);
            return quote! {
                if let crate::cst::#children::#variant(#current_node) = #current_node {
                    #base_matcher
                }
            };
        }
    }
    fn group_children<'b>(
        &self,
        node: &ts_query::NamedNode<'b>,
        first_node: &ts_query::NamedNodeChildren<'_>,
        mut name_value: Option<TokenStream>,
        current_node: &Ident,
    ) -> (Option<TokenStream>, Vec<ts_query::NamedNodeChildren<'b>>) {
        let mut prev = first_node.clone();
        let mut remaining_nodes = Vec::new();
        for child in node.children().into_iter().skip(1) {
            if child.kind() == "capture" {
                if child.source() == "@name" {
                    log::info!(
                        "Found @name! prev: {:#?}, {:#?}",
                        prev.source(),
                        prev.kind()
                    );
                    match prev {
                        ts_query::NamedNodeChildren::FieldDefinition(field) => {
                            let field_name = field
                                .name
                                .iter()
                                .filter(|c| c.is_named())
                                .map(|c| format_ident!("{}", c.source()))
                                .next()
                                .unwrap();
                            name_value = Some(quote! {
                                #current_node.#field_name.source()
                            });
                        }
                        ts_query::NamedNodeChildren::Identifier(named) => {
                            log::info!(
                                "Found @name! prev: {:#?}, {:#?}",
                                named.source(),
                                named.kind()
                            );
                            name_value = Some(quote! {
                                #current_node.source()
                            });
                        }
                        _ => panic!(
                            "Unexpected prev: {:#?}, source: {:#?}",
                            prev.kind(),
                            prev.source()
                        ),
                    }
                    break;
                }
                continue;
            }
            prev = child.clone();
            remaining_nodes.push(child);
        }
        (name_value, remaining_nodes)
    }
    fn get_matcher_for_named_node(
        &self,
        node: &ts_query::NamedNode,
        struct_name: &str,
        current_node: &Ident,
        name_value: Option<TokenStream>,
    ) -> TokenStream {
        let mut matchers = TokenStream::new();
        let first_node = node.children().into_iter().next().unwrap();
        let (name_value, remaining_nodes) =
            self.group_children(node, &first_node, name_value, current_node);
        if remaining_nodes.len() == 0 {
            log::info!("single node, {}", first_node.source());
            return self.get_matcher_for_definition(
                struct_name,
                first_node.into(),
                current_node,
                name_value,
            );
        }

        let name_node = self.state.get_node_for_raw_name(&first_node.source());
        if let Some(name_node) = name_node {
            let target_name = name_node.normalize_name();
            let matcher = self._get_matcher_for_named_node(
                struct_name,
                &target_name,
                name_node.kind(),
                current_node,
                remaining_nodes,
                name_value,
            );
            matchers.extend_one(matcher);
        } else {
            let subenum = self.state.get_subenum_variants(&first_node.source());
            log::info!(
                "subenum {} with {} variants",
                first_node.source(),
                subenum.len()
            );
            for variant in subenum {
                if variant.normalize_name() == "Comment" {
                    continue;
                }
                let matcher = self._get_matcher_for_named_node(
                    struct_name,
                    &variant.normalize_name(),
                    variant.kind(),
                    current_node,
                    remaining_nodes.clone(),
                    name_value.clone(),
                );
                matchers.extend_one(matcher);
            }
        }
        quote! {
            #matchers
        }
    }
    fn get_default_matcher(&self, name_value: Option<TokenStream>) -> TokenStream {
        let to_append = self.executor_id();
        let name_value = name_value
            .ok_or(format!("No name value found for: {}", self.node().source()))
            .unwrap();
        quote! {
            self.#to_append.insert(#name_value,node.clone());
        }
    }
    fn get_matcher_for_identifier(
        &self,
        identifier: &ts_query::Identifier,
        struct_name: &str,
        current_node: &Ident,
        name_value: Option<TokenStream>,
    ) -> TokenStream {
        // We have 2 nodes, the parent node and the identifier node
        let to_append = self.get_default_matcher(name_value);
        let children;
        // Case 1: The identifier is the same as the struct name (IE: we know this is the corrent node)
        if normalize_type_name(&identifier.source(), true) == struct_name {
            return to_append;
        }
        // Case 2: We have a node for the parent struct
        if let Some(node) = self.state.get_node_for_struct_name(struct_name) {
            children = format_ident!("{}Children", struct_name);
            // When there is only 1 possible child, we can use the default matcher
            if node.children_struct_name() != children.to_string() {
                return to_append;
            }
        } else {
            // Case 3: This is a subenum
            // If this is a field, we may be dealing with multiple types and can't operate over all of them
            return to_append; // TODO: Handle this case
        }
        let struct_name = format_ident!("{}", normalize_type_name(&identifier.source(), true));
        quote! {
            if let crate::cst::#children::#struct_name(child) = #current_node {
                #to_append
            }

        }
    }
    fn get_matcher_for_definition(
        &self,
        struct_name: &str,
        node: ts_query::NodeTypes,
        current_node: &Ident,
        name_value: Option<TokenStream>,
    ) -> TokenStream {
        if !node.is_named() {
            return self.get_default_matcher(name_value);
        }
        match node {
            ts_query::NodeTypes::FieldDefinition(field) => {
                self.get_matcher_for_field(&field, struct_name, current_node, name_value)
            }
            ts_query::NodeTypes::Capture(named) => {
                info!("Capture: {:#?}", named.source());
                quote! {}
            }
            ts_query::NodeTypes::NamedNode(named) => {
                self.get_matcher_for_named_node(&named, struct_name, current_node, name_value)
            }
            ts_query::NodeTypes::Comment(_) => {
                quote! {}
            }
            ts_query::NodeTypes::List(subenum) => {
                for child in subenum.children() {
                    let result = self.get_matcher_for_definition(
                        struct_name,
                        child.into(),
                        current_node,
                        name_value.clone(),
                    );
                    // Currently just returns the first child
                    return result; // TODO: properly handle list
                }
                quote! {}
            }
            ts_query::NodeTypes::Grouping(grouping) => {
                self.get_matchers_for_grouping(&grouping, struct_name, current_node, name_value)
            }
            ts_query::NodeTypes::Identifier(identifier) => {
                self.get_matcher_for_identifier(&identifier, struct_name, current_node, name_value)
            }
            unhandled => {
                log::warn!(
                    "Unhandled definition in language {}: {:#?}, {:#?}",
                    self.language.name(),
                    unhandled.kind(),
                    unhandled.source()
                );
                self.get_default_matcher(name_value)
            }
        }
    }

    pub fn matcher(&self, struct_name: &str) -> TokenStream {
        let node = self.state.get_node_for_struct_name(struct_name);
        let kind = if let Some(node) = node {
            node.kind()
        } else {
            struct_name
        };
        let starting_node = format_ident!("node");
        let (name_value, remaining_nodes) = self.group_children(
            &self.node(),
            &self.node().children().into_iter().next().unwrap(),
            None,
            &starting_node,
        );
        return self._get_matcher_for_named_node(
            struct_name,
            &struct_name,
            kind,
            &starting_node,
            remaining_nodes,
            name_value,
        );
    }
}

pub trait HasQuery {
    fn queries<'a, 'db: 'a>(&'a self, db: &'db dyn salsa::Database) -> BTreeMap<String, Query<'a>>;
    fn queries_with_prefix<'a, 'db: 'a>(
        &'a self,
        db: &'db dyn salsa::Database,
        prefix: &str,
    ) -> BTreeMap<String, Vec<Query<'a>>> {
        let mut queries = BTreeMap::new();
        for (name, query) in self.queries(db).into_iter() {
            if name.starts_with(prefix) {
                let new_name = name.split(".").last().unwrap();
                queries
                    .entry(new_name.to_string())
                    .or_insert(Vec::new())
                    .push(query);
            }
        }
        queries
    }
}
impl HasQuery for Language {
    fn queries<'a, 'db: 'a>(&'a self, db: &'db dyn salsa::Database) -> BTreeMap<String, Query<'a>> {
        Query::from_queries(db, &self.tag_query, self)
    }
}
#[cfg(test)]
mod tests {
    use codegen_sdk_common::language::ts_query;
    use codegen_sdk_cst::CSTDatabase;

    use super::*;
    #[test]
    fn test_query_basic() {
        let database = CSTDatabase::default();
        let language = &ts_query::Query;
        let queries = Query::from_queries(&database, "(abc) @definition.abc", language);
        assert!(queries.len() > 0);
    }
}
