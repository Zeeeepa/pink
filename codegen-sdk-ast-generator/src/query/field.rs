use proc_macro2::Span;
use syn::parse_quote;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Field {
    pub name: String,
    pub kind: String,
    pub is_optional: bool,
    pub is_multiple: bool,
}
impl Field {
    pub fn as_syn_field(&self) -> syn::Field {
        let name_ident = syn::Ident::new(&self.name, Span::call_site());
        let type_name = syn::Ident::new(&self.kind, Span::call_site());
        parse_quote!(
            #[tracked]
            #[return_ref]
            pub #name_ident: crate::cst::#type_name<'db>
        )
    }
}
