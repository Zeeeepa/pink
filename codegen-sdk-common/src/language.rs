use tree_sitter::{LanguageError, Parser};

pub struct Language {
    pub name: &'static str,
    pub node_types: &'static str,
    pub file_extensions: &'static [&'static str],
    pub tree_sitter_language: tree_sitter::Language,
}
impl Language {
    pub fn parse_tree_sitter(&self, content: &str) -> Result<tree_sitter::Tree, LanguageError> {
        let mut parser = Parser::new();
        parser.set_language(&self.tree_sitter_language)?;
        let tree = parser.parse(content, None).unwrap();
        Ok(tree)
    }
}
#[cfg(feature = "typescript")]
pub mod javascript;
#[cfg(feature = "typescript")]
pub mod jsx;
#[cfg(feature = "python")]
pub mod python;
#[cfg(feature = "typescript")]
pub mod tsx;
#[cfg(feature = "typescript")]
pub mod typescript;
lazy_static! {
    pub static ref LANGUAGES: Vec<&'static Language> = vec![
        #[cfg(feature = "python")]
        &python::Python,
        #[cfg(feature = "typescript")]
        &typescript::Typescript,
        #[cfg(feature = "typescript")]
        &tsx::TSX,
        #[cfg(feature = "typescript")]
        &jsx::JSX,
        #[cfg(feature = "typescript")]
        &javascript::Javascript,
    ];
}
