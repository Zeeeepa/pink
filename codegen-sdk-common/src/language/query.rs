use super::Language;
lazy_static! {
    pub static ref Query: Language = Language::new(
        "query",
        "Query",
        tree_sitter_query::NODE_TYPES,
        &[],
        tree_sitter_query::LANGUAGE.into(),
        "",
    )
    .unwrap();
}
