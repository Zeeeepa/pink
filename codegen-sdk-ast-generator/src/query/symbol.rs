use pluralizer::pluralize;
use proc_macro2::Span;
use syn::parse_quote_spanned;

use crate::query::field::Field;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Symbol {
    pub name: String,
    pub type_name: String,
    pub category: String,
    pub subcategory: String,
    pub language_struct: String,
    pub fields: Vec<Field>,
}
impl Symbol {
    pub fn as_syn_struct(&self) -> Vec<syn::Stmt> {
        let span = Span::call_site();
        let variant = syn::Ident::new(&self.name, span);
        let type_name = syn::Ident::new(&self.type_name, span);
        let language_struct = syn::Ident::new(&self.language_struct, span);
        let fields = self
            .fields
            .iter()
            .map(|field| field.as_syn_field())
            .collect::<Vec<_>>();
        let getters = self
            .fields
            .iter()
            .map(|field| field.getter())
            .collect::<Vec<_>>();
        parse_quote_spanned! {
            span =>
            #[salsa::tracked]
            pub struct #variant<'db> {
                #[id]
                _fully_qualified_name: codegen_sdk_resolution::FullyQualifiedName,
                #[id]
                node_id: indextree::NodeId,
                // #[tracked]
                // #[return_ref]
                // pub node: crate::cst::#type_name<'db>,
                #(#fields),*
            }
            impl<'db> #variant<'db> {
                pub fn node(&self, db: &'db dyn codegen_sdk_resolution::Db) -> &'db crate::cst::#type_name<'db> {
                    let file = self.file(db);
                    let tree = file.tree(db);
                    tree.get(&self.node_id(db)).unwrap().as_ref().try_into().unwrap()
                }
                #(#getters)*
            }
            impl<'db> codegen_sdk_resolution::HasFile<'db> for #variant<'db> {
                type File<'db1> = #language_struct<'db1>;
                fn file(&self, db: &'db dyn codegen_sdk_resolution::Db) -> &'db Self::File<'db> {
                    let path = self._fully_qualified_name(db).path(db);
                    parse(db, path)
                }
                fn root_path(&self, db: &'db dyn codegen_sdk_resolution::Db) -> PathBuf {
                    self.node(db).id().root(db).path(db)
                }
            }
            impl<'db> codegen_sdk_resolution::HasId<'db> for #variant<'db> {
                fn fully_qualified_name(&self, db: &'db dyn salsa::Database) -> codegen_sdk_resolution::FullyQualifiedName {
                    self._fully_qualified_name(db)
                }
            }
        }
    }
    pub fn py_file_getter(&self) -> syn::Stmt {
        let span = Span::call_site();
        let method_name = syn::Ident::new(&pluralize(self.name.as_str(), 2, false), span);
        let category = syn::Ident::new(&self.category, span);
        let type_name = syn::Ident::new(&self.name, span);
        let subcategory = syn::Ident::new(&self.subcategory, span);
        parse_quote_spanned! {
            span =>
            #[getter]
            pub fn #method_name(&self, py: Python<'_>) -> PyResult<Vec<#type_name>> {
                let file = self.file(py)?;
                let db = self.codebase.get(py).db();
                let category = file.#category(db);
                let subcategory = category.#subcategory(db);
                let nodes = subcategory.values().flatten().map(|node| #type_name::new(node.fully_qualified_name(), self.codebase.clone())).collect();
                Ok(nodes)
            }
        }
    }
}
