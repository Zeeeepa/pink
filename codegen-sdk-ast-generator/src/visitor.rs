use std::collections::{BTreeMap, BTreeSet};

use codegen_sdk_common::Language;
use convert_case::{Case, Casing};
use proc_macro2::{Span, TokenStream};
use quote::{format_ident, quote};
use syn::parse_quote_spanned;

use super::query::Query;
use crate::query::HasQuery;
pub fn generate_visitor<'db>(
    db: &'db dyn salsa::Database,
    language: &Language,
    name: &str,
) -> TokenStream {
    log::info!(
        "Generating visitor for language: {} for {}",
        language.name(),
        name
    );
    let raw_queries = language.queries_with_prefix(db, &format!("{}", name));
    let queries: Vec<&Query> = raw_queries.values().flatten().collect();
    let mut names = Vec::new();
    let mut types = Vec::new();
    let mut variants = BTreeSet::new();
    let mut enter_methods = BTreeMap::new();
    for query in queries {
        names.push(query.executor_id());
        types.push(format_ident!("{}", query.struct_name()));
        for variant in query.struct_variants() {
            variants.insert(format_ident!("{}", variant));
            enter_methods
                .entry(variant)
                .or_insert(Vec::new())
                .push(query);
        }
    }
    let mut methods: Vec<syn::Arm> = Vec::new();
    for (variant, queries) in enter_methods {
        let mut matchers = TokenStream::new();
        let struct_name = format_ident!("{}", variant);
        for query in queries {
            matchers.extend_one(query.matcher(&variant));
        }
        let span = Span::mixed_site();
        methods.push(parse_quote_spanned! { span =>
            crate::cst::NodeTypes::#struct_name(node) => {
                #matchers
            }
        });
    }
    let maps = quote! {
        #(
            let mut #names: BTreeMap<String,Vec<indextree::NodeId>> = BTreeMap::new();
        )*
    };
    let constructor = quote! {
        Self::new(db, #(#names),*)

    };
    let name = format_ident!("{}s", name.to_case(Case::Pascal));
    let output_constructor = quote! {
        pub fn visit(db: &'db dyn salsa::Database, root: &'db crate::cst::Parsed<'db>) -> Self {
            #maps
            let tree = root.tree(db);
            for (node, id) in tree.descendants(&root.program(db)) {
                match node {
                    #(#methods,)*
                    _ => {}
                }
            }
            #constructor
        }
        pub fn default(db: &'db dyn salsa::Database) -> Self {
            #maps
            #constructor
        }
    };
    let underscored_names = names
        .iter()
        .map(|name| format_ident!("_{}", name))
        .collect::<Vec<_>>();
    quote! {
        // Three lifetimes:
        // db: the lifetime of the database
        // db1: the lifetime of the visitor executing per-node
        // db2: the lifetime of the references held by the visitor
        #[salsa::tracked]
        pub struct #name<'db> {
            #(
                #[return_ref]
                pub #underscored_names: BTreeMap<String, Vec<indextree::NodeId>>,
            )*
        }
        impl<'db> #name<'db> {
            #output_constructor
            #(
                pub fn #names(&self, db: &'db dyn salsa::Database, tree: &'db codegen_sdk_common::tree::Tree<crate::cst::NodeTypes<'db>>) -> BTreeMap<String, Vec<&'db crate::cst::#types<'db>>> {
                    self.#underscored_names(db).iter().map(|(k, v)|
                    (k.clone(), v.iter().map(|id| tree.get(id).unwrap().as_ref().try_into().unwrap()).collect())).collect()
                }
            )*
        }
    }
}

#[cfg(all(test))]
mod tests {
    use codegen_sdk_common::language::{python::Python, typescript::Typescript};
    use rstest::rstest;

    use super::*;

    #[test_log::test(rstest)]
    #[case::typescript(&Typescript)]
    #[case::python(&Python)]
    fn test_generate_visitor(#[case] language: &Language) {
        let db = codegen_sdk_cst::CSTDatabase::default();
        let visitor = generate_visitor(&db, language, "definition");
        insta::assert_snapshot!(
            format!("{}", language.name()),
            codegen_sdk_common::generator::format_code_string(&visitor.to_string()).unwrap()
        );
    }
}
