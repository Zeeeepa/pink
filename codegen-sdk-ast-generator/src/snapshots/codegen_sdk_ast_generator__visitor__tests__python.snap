---
source: codegen-sdk-ast-generator/src/visitor.rs
expression: "codegen_sdk_common::generator::format_code_string(&visitor.to_string()).unwrap()"
---
#[salsa::tracked]
pub struct Class<'db> {
    #[id]
    node_id: indextree::NodeId,
    #[tracked]
    #[return_ref]
    pub node: crate::cst::ClassDefinition<'db>,
    pub name: crate::cst::Identifier<'db>,
}
#[salsa::tracked]
pub struct Constant<'db> {
    #[id]
    node_id: indextree::NodeId,
    #[tracked]
    #[return_ref]
    pub node: crate::cst::Module<'db>,
    pub name: crate::cst::Identifier<'db>,
}
#[salsa::tracked]
pub struct Function<'db> {
    #[id]
    node_id: indextree::NodeId,
    #[tracked]
    #[return_ref]
    pub node: crate::cst::FunctionDefinition<'db>,
    pub name: crate::cst::Identifier<'db>,
}
#[salsa::tracked]
pub struct Import<'db> {
    #[id]
    node_id: indextree::NodeId,
    #[tracked]
    #[return_ref]
    pub node: crate::cst::ImportFromStatement<'db>,
    pub module: crate::cst::NodeTypes<'db>,
    pub name: crate::cst::NodeTypes<'db>,
}
#[derive(Debug, Clone, PartialEq, Eq, Hash, salsa::Update)]
pub enum Symbol<'db> {
    Class(Class<'db>),
    Constant(Constant<'db>),
    Function(Function<'db>),
    Import(Import<'db>),
}
#[salsa::tracked]
pub struct Definitions<'db> {
    #[return_ref]
    pub classes: BTreeMap<String, Vec<Class<'db>>>,
    #[return_ref]
    pub constants: BTreeMap<String, Vec<Constant<'db>>>,
    #[return_ref]
    pub functions: BTreeMap<String, Vec<Function<'db>>>,
    #[return_ref]
    pub imports: BTreeMap<String, Vec<Import<'db>>>,
}
impl<'db> Definitions<'db> {
    pub fn visit(
        db: &'db dyn salsa::Database,
        root: &'db crate::cst::Parsed<'db>,
    ) -> Self {
        let mut classes: BTreeMap<String, Vec<Class<'db>>> = BTreeMap::new();
        let mut constants: BTreeMap<String, Vec<Constant<'db>>> = BTreeMap::new();
        let mut functions: BTreeMap<String, Vec<Function<'db>>> = BTreeMap::new();
        let mut imports: BTreeMap<String, Vec<Import<'db>>> = BTreeMap::new();
        let tree = root.tree(db);
        for (node, id) in tree.descendants(&root.program(db)) {
            match node {
                crate::cst::NodeTypes::ClassDefinition(node) => {
                    ///Code for query: (class_definition   name: (identifier) @name) @definition.class
                    let name = node.name(tree);
                    let symbol = Class::new(db, id, node.clone(), name.clone());
                    classes.entry(name.source()).or_default().push(symbol);
                }
                crate::cst::NodeTypes::FunctionDefinition(node) => {
                    ///Code for query: (function_definition   name: (identifier) @name) @definition.function
                    let name = node.name(tree);
                    let symbol = Function::new(db, id, node.clone(), name.clone());
                    functions.entry(name.source()).or_default().push(symbol);
                }
                crate::cst::NodeTypes::ImportFromStatement(node) => {
                    ///Code for query: (import_from_statement module_name: (_) @module name: (_) @name) @definition.import
                    let module_name = node.module_name(tree);
                    for name in node.name(tree) {
                        let symbol = Import::new(
                            db,
                            id,
                            node.clone(),
                            module_name.clone(),
                            name.clone(),
                        );
                        imports.entry(name.source()).or_default().push(symbol);
                    }
                }
                crate::cst::NodeTypes::Module(node) => {
                    ///Code for query: (module (expression_statement (assignment left: (identifier) @name) @definition.constant))
                    for child in node.children(tree) {
                        if let crate::cst::ModuleChildrenRef::ExpressionStatement(
                            child,
                        ) = child {
                            ///Code for query: (module (expression_statement (assignment left: (identifier) @name) @definition.constant))
                            for child in child.children(tree) {
                                if let crate::cst::ExpressionStatementChildrenRef::Assignment(
                                    child,
                                ) = child {
                                    ///Code for query: (module (expression_statement (assignment left: (identifier) @name) @definition.constant))
                                    let left = child.left(tree);
                                    if let crate::cst::AssignmentLeftRef::Identifier(left) = left {
                                        let symbol = Constant::new(
                                            db,
                                            id,
                                            node.clone(),
                                            left.clone(),
                                        );
                                        constants.entry(left.source()).or_default().push(symbol);
                                    }
                                }
                                break;
                            }
                        }
                        break;
                    }
                }
                _ => {}
            }
        }
        Self::new(db, classes, constants, functions, imports)
    }
    pub fn default(db: &'db dyn salsa::Database) -> Self {
        let mut classes: BTreeMap<String, Vec<Class<'db>>> = BTreeMap::new();
        let mut constants: BTreeMap<String, Vec<Constant<'db>>> = BTreeMap::new();
        let mut functions: BTreeMap<String, Vec<Function<'db>>> = BTreeMap::new();
        let mut imports: BTreeMap<String, Vec<Import<'db>>> = BTreeMap::new();
        Self::new(db, classes, constants, functions, imports)
    }
}
